usr/lib/common/asn1.c

```diff
@@ -4215,12 +4215,28 @@ CK_RV ber_decode_IBM_DilithiumPrivateKey(CK_BYTE *data,
         offset += field_len;
     }
 
-    /* t1 */
-    rc = ber_decode_BIT_STRING(buf + offset, &tmp, &len, &field_len);
-    if (rc != CKR_OK) {
-        TRACE_DEVEL("ber_decode_BIT_STRING of (t1) failed\n");
-        goto cleanup;
-    } else {
+    /* t1 (optional, within choice) */
+    if (offset < buf_len) {
+        rc = ber_decode_CHOICE(buf + offset, &tmp, &len, &field_len, &option);
+        if (rc != CKR_OK) {
+            TRACE_DEVEL("ber_decode_BIT_STRING of (t1) failed\n");
+            goto cleanup;
+        }
+
+        if (option != 0x00) {
+            TRACE_DEVEL("ber_decode_CHOICE returned invalid option %ld\n",
+                        option);
+            goto cleanup;
+        }
+
+        offset += field_len - len;
+
+        rc = ber_decode_BIT_STRING(buf + offset, &tmp, &len, &field_len);
+        if (rc != CKR_OK) {
+            TRACE_DEVEL("ber_decode_BIT_STRING of (t1) failed\n");
```

Comment:
What happens if option == 0, but there is no (t1) value given? Will field_len be just 0 and the build_attribute below creates a "dummy" attr with a length of 0? Should we probably don't add the attr at all if (t1) is not given?

Reply:
Well, the same question applies to all the key components, even those that are not optional by design. If their length would be zero we would add a zero length attribute (build_attribute can deal with len = 0).

That would anyway be an invalid encoding of an Dilithium key. 
So: garbage in, garbage out :-)